### ISSUE TITLE
Nginx not connecting to upstream hosts after host upgrades

### ISSUE THREAD

**Issue by steveabbitt** [rancher/rancher#6375]
**Rancher Version:**
v1.1.4
**Docker Version:**
v1.12.1
**OS and where are the hosts located? (cloud, bare metal, etc):**
Ubuntu 16.04.1 on AWS EC2
**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
Single node Rancher
**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)**
Cattle
**Steps to Reproduce:**
Build Nginx container that points to other upstream containers hosting web app. Upgrade any upstream app container.
**Results:** 
Upgraded host returns an Nginx 502 Bad Gateway after upstream app upgrade
**Expected:**
Nginx routes to upgraded app

Everything works fine after setting up Nginx, but after upgrading any of the app containers, I get the same issue. Saw some other issue threads that look related with mentions that the problem was "being worked on" months ago, but they were closed or abandoned with no resolution that I could see.

**Comment by ebr** [comment]
I believe this has to do with nginx caching the upstream IPs. The way we're working around this is by running a dnsmasq daemon inside the nginx container, and nginx using that as a resolver. This way any changes to the Rancher services are resolved transparently to nginx.

`dnsmasq.conf`:
```
interface=lo
listen-address=127.0.0.1
no-dhcp-interface=lo
strict-order
```

`nginx.conf`:
```
...
	resolver 127.0.0.1 valid=1s ipv6=off;
...
```
`Dockerfile` to build nginx image (just the relevant parts):
```
FROM nginx:mainline-alpine
RUN apk add --update dnsmasq
COPY dnsmasq.conf /etc/dnsmasq.conf
```

**Comment by steveabbitt** [comment]
This solution did not fix the issue. Still having the same problem.

**Comment by aemneina** [comment]
@steveabbitt can you share your nginx configuration? how are you configuring the backend services, through IP addresses?

**Comment by steveabbitt** [comment]
The backend services are configured using the Docker container names. I have about 20 nearly identical container apps that are configured. The following nginx.conf is pared down to only show one of them. I have also tried the 'resolver' line in the main http block with no success.

```
daemon off;
error_log /dev/stdout warn;
worker_processes  4;
events { worker_connections 1024; }

http {
  proxy_buffering    off;
  proxy_set_header   X-Real-IP $remote_addr;
  proxy_set_header   X-Scheme $scheme;
  proxy_set_header   X-Forwarded-For $proxy_add_x_forwarded_for;
  proxy_set_header   Host $http_host;
  proxy_http_version 1.1;
  access_log /dev/stdout;
  server_tokens off;
  add_header X-Frame-Options SAMEORIGIN;
  include /etc/nginx/conf/botlimit.conf;

 upstream test-up {
  server test:9000;
 }
 server {
  listen 80;
  return 301 https://$host$request_uri;
 }
 server {
  listen 8080;
  keepalive_timeout 70;
  client_max_body_size 50M;
  client_body_timeout 5s;
  client_header_timeout 5s;
  add_header Strict-Transport-Security "max-age=31536000; includeSubDomains";
  location / {
   proxy_pass http://test-up;
   include /etc/nginx/conf/blockbots.conf;
   resolver 127.0.0.1 valid=1s ipv6=off;
  }
 }
}
```

**Comment by ebr** [comment]
@steveabbitt You should read up on how Nginx caches the `upstream` directive. 

https://www.ruby-forum.com/topic/4407628
https://github.com/GUI/nginx-upstream-dynamic-servers (see useful links at the end of README)
https://www.nginx.com/blog/dns-service-discovery-nginx-plus/

Dynamic resolution of upstreams is only available in Nginx Plus 💰 💰 . The default behaviour is to resolve any DNS names once when the server starts, and only continue using the cached values.

TL;DR: do not use the `upstream` directive for this, or use a 3rd party module.

If you're running `dnsmasq` in the same container as nginx, as suggested, you'd need to:
1. add `resolver 127.0.0.1 valid=1s ipv6=off;` directive in the main block of your config, 
2. in your `server` block, just `proxy_pass http://test:9000;`.

**Comment by abc34** [comment]
Add [https://tenzer.dk/nginx-with-dynamic-upstreams/](url)

**Comment by steveabbitt** [comment]
Thank you for all the responses. I fixed it by configuring the app containers to retain their IP addresses when upgrading/moving. Nginx can cache them all it wants now. :)

**Comment by ebr** [comment]
@steveabbitt yep, that's also a solution! 😉  
(although, what happens if you scale up the upstreams, and some new containers are started whose IPs nginx isn't yet aware of?)

**Comment by steveabbitt** [comment]
@ebr I thought about that, but in this particular situation that won't be an issue. Stable deploys of the existing containers are mission critical during business hours. Any additions or scaling can be done at times when brief downtime is not an issue while Nginx restarts.

**Comment by aemneina** [comment]
awesome, closing this one out.
